docs/src/design/changing_the_primal.md

```diff
@@ -0,0 +1,435 @@
+  1.255 ns (0 allocations: 0 bytes)
+
+julia> @btime 1/(1+ex) setup=(ex=exp(rand()));
+  1.256 ns (0 allocations: 0 bytes)
+
+julia> 5.367 + 1.255 + 1.256
+7.878
+```
+
+So we are talking about a 30-40%[^4] speed-up from these optimizations.
+It's faster to  compute `sin` and `cos` at the same time via `sincos` than it is to compute them one after the other.
+And it is faster to reuse the `exp(x)` in computing `σ(x)` and `σ(-x)`.
+How can we incorporate this insight into our system?
+We know we can compute both of these in the primal, because they only depend on `x` --- we don't need to know `ȳ`.
+But there is nowhere to put it that is accessible both to the primal pass and the gradient pass code.
+
+
+What if we introduced some variable called `intermediates` that is also recorded onto the tape during the primal pass?
+We would need to be able to modify the primal pass to do this, so that we can actually put the data into the `intermediates`.
+So we will introduce a function: `augmented_primal`, that will return the primal output plus the `intermediates` that we want to reuse in the gradient pass.
+Then we will make our AD system replace calls to the primal with calls to the `augmented_primal` of the primal function and take care of all the bookkeeping.
+So that would look like:
+```julia
+y = f(x)  # primal program
+y, intermediates = augmented_primal(f, x)
+x̄ = pullback_at(f, x, y, ȳ, intermediates)
+```
+
+Let's try writing this now for `sin`
+```julia
+function augmented_primal(::typeof(sin), x)
+  y, cx = sincos(x)
+  return y, (; cx=cx)  # use a NamedTuple for the intermediates
+end
+
+pullback_at(::typeof(sin), x, y, ȳ, intermediates) = ȳ * intermediates.cx
+```
+and for the logistic sigmoid:
+```julia
+function augmented_primal(::typeof(σ), x)
+  ex = exp(x)
+  y = ex/(1 + ex)
+  return y, (; ex=ex)  # use a NamedTuple for the intermediates
+end
+
+pullback_at(::typeof(σ), x, y, ȳ, intermediates) = ȳ * y /(1 + intermediates.ex)
+```
+
+
+Cool!
+That lets us do what we wanted.
+We net decreased the time it takes to run the primal and gradient passes.
+We have now demonstrated the title question of why we needed to be able to modify the primal pass.
+We will go into that more later, and have some more examples of use.
+But first let's continue to see how we go from that `augmented_primal` to `pullback_at` to [`rrule`](@ref).
+
+
+One thing we notice when looking at `pullback_at` is it really is starting to have a lot of arguments.
+It had a fair few already, and now we are adding `intermediates` as well.
+Not to mention this is a fairly simple function, only 1 input, no keyword arguments.
+Furthermore, we don't even use all of them all the time.
+The original new code for pulling back `sin` no longer needs the `x`, and it never needed `y` (though `sigmoid` does).
+Having all this extra stuff means that the `pullback_at` function signature is long, and we are putting a bunch of extra stuff on the tape, using more memory.
```

Comment:
```suggestion
The function signature for `pullback_at` was already rather long with the primal inputs and outputs and the sensitivity.
Now it is even longer since we added `intermediates`.
Further, storing all all those things on the tape is using up extra memory.
```